## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(f"The maximum quantity ordered is: {max(df['qty'])}")
print(f"The average quantity ordered is: {df['qty'].mean()}")
print(f"The minimum quantity ordered is: {min(df['qty'])} \n")

print(f"The maximum unit price ordered is: {max(df['unit_price'])}")
print(f"The average unit price ordered is: {df['unit_price'].mean()}")
print(f"The minimum unit price ordered is: {min(df['unit_price'])} \n")

print(f"The maximum unit cost ordered is: {max(df['unit_cost'])}")
print(f"The average unit price ordered is: {df['unit_cost'].mean()}")
print(f"The minimum unit cost ordered is: {min(df['unit_cost'])} \n")

print(f"The maximum unit weight ordered is: {max(df['unit_weight'])}")
print(f"The average unit price ordered is: {df['unit_weight'].mean()}")
print(f"The minimum unit weight ordered is: {min(df['unit_weight'])} \n")

print("Below is a table of the number of different jobs present: ")
print(df['job'].value_counts())

print("Below is a table of the different catecories present: ")
print(df["category"].value_counts())

print("Below is a table of the different subcatecories present: ")
print(df["subcategory"].value_counts())

The maximum quantity ordered is: 3958244
The average quantity ordered is: 570.2646095279928
The minimum quantity ordered is: 0 

The maximum unit price ordered is: 1396.23
The average unit price ordered is: 136.26720694009774
The minimum unit price ordered is: 0.01 

The maximum unit cost ordered is: 846.27
The average unit price ordered is: 99.4460727685353
The minimum unit cost ordered is: 0.01 

The maximum unit weight ordered is: 46.43
The average unit price ordered is: 5.0041162905616865
The minimum unit weight ordered is: 0.0 

Below is a table of the number of different jobs present: 
job
Osteopath                             631
Toxicologist                          546
Architectural technologist            430
Librarian, academic                   389
Teacher, special educational needs    388
                                     ... 
Copywriter, advertising                12
Midwife                                10
Insurance risk surveyor                 9
Runner, broadcastin

In [8]:
# What three item categories had the most entries?
print("The three item categories with the most entries are: ")
print(df["category"].value_counts()[:3])

The three item categories with the most entries are: 
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [17]:
# For the category with the most entries, which subcategory had the most entries?

highest_cat = df['category'].value_counts().index[0]

highest_cat_dataframe = df.loc[df['category'] == highest_cat, :]
highest_subcat = highest_cat_dataframe['subcategory'].value_counts().index[0]
print(f"For the category '{highest_cat}' which has the most entries, the subcategory with the most entries is: {highest_subcat}")


For the category 'consumables' which has the most entries, the subcategory with the most entries is: bathroom supplies


In [19]:
# Which five clients had the most entries in the data?
print("The five clients with the most entries in the data are: ")
print(df['client_id'].value_counts()[:5])

The five clients with the most entries in the data are: 
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [21]:
# Store the client ids of those top 5 clients in a list.
top_five = list(df['client_id'].value_counts().index[:5])
top_five

[33615, 66037, 46820, 38378, 24741]

In [28]:
# How many total units (the qty column) did the client with the most entries order order?
# Homework assumption, the total units means every last order made by said client, i.e. a summation of all the client's orders
max_qty = df.loc[(df['client_id'] == top_five[0]), "qty"].sum()
print(f"The total units ordered by the client with the most entries is: {max_qty}")


The total units ordered by the client with the most entries is: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [31]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
# For homework the assumption is that the column is meant to be added to the dataframe
df['subtotal'] = df["unit_price"] * df["qty"]
# Show that the dataframe now has the same columns as before along with the new column
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal'],
      dtype='object')

In [35]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
#First find the total weight
total_weight = df['qty'] * df['unit_weight']
#Get a boolean of where the total weight is above 50 then multiply by the cost of $7
total_weight_above = (total_weight > 50) * 7
#Get a boolean of where the total weight is 50 or below then multiply by the cost of $10
total_weight_below = (total_weight <= 50) * 10
#Use the resultant shipping cost areas to generate the total cost (as the areas where the weight is incorrect, the value will be zero)
total_weight_above = total_weight * total_weight_above
total_weight_below = total_weight * total_weight_below
#Then add the two boolean based vectors to determine the shipping price
df['shipping_price'] = total_weight_above + total_weight_below
#Print the first three of the quantity, unit_weight, and shipping price as a sanity check
print(df[['qty', 'unit_weight', 'shipping_price']][:3])


   qty  unit_weight  shipping_price
0  105         7.50         5512.50
1   21         1.49          312.90
2   39         1.68          458.64


In [37]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
#Assumption here is that sales tax is a constant so it is put in all caps
SALES_TAX = 0.0925
df['total_price'] = (df['subtotal'] + df['shipping_price']) * SALES_TAX + (df['subtotal'] + df['shipping_price'])
#Print the first three of the subtotal, shipping price, and total price as a sanity check
print(df[['subtotal','shipping_price' ,'total_price']][:3])

    subtotal  shipping_price    total_price
0  115164.00         5512.50  131839.076250
1     523.95          312.90     914.258625
2     527.28          458.64    1077.117600


In [38]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
#Assumption here is that the sales tax is not part of the cost as it is paid by the client.
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
#Sanity check
print(df[['qty', 'unit_cost', 'shipping_price', 'line_cost']][:3])


   qty  unit_cost  shipping_price  line_cost
0  105     762.71         5512.50   85597.05
1   21      15.09          312.90     629.79
2   39       7.86          458.64     765.18


In [39]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
print(df['profit'][:3])


0    46242.026250
1      284.468625
2      311.937600
Name: profit, dtype: float64


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [69]:
# Check your work using the totals above
order_check_first = df.loc[df['order_id'] == 2742071]
print(f"The order ID 2742071 has a total price of: {order_check_first['total_price'].sum().round(2)}")


order_check_second = df.loc[df['order_id'] == 2173913]
print(f"The order ID 2173913 has a total price of: {round(order_check_second['total_price'].sum(), 2)}")

order_check_third = df.loc[df['order_id'] == 6128929]
print(f"The order ID 6128929 has a total price of: {round(order_check_third['total_price'].sum(), 2)}")

The order ID 2742071 has a total price of: 152811.9
The order ID 2173913 has a total price of: 162388.72
The order ID 6128929 has a total price of: 923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [72]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
#Assumptions: that the spending includes sales tax, shipping, etc, i.e. that the total price is the client's spending
for big_cat in top_five:
    big_monies = df.loc[df['client_id'] == big_cat, 'total_price'].sum()
    print(f"The top spender {big_cat} spent ${round(big_monies, 2)}")


The top spender 33615 spent $8377308.57
The top spender 66037 spent $10259514.8
The top spender 46820 spent $9743794.32
The top spender 38378 spent $12906550.88
The top spender 24741 spent $82268891.98


In [136]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Got homework help from William Pitera who reminded me about the groupby function! 
top_five_df = df[df['client_id'].isin(top_five)].groupby('client_id').agg({
    'qty':'sum',
    'shipping_price':'sum',
    'total_price':'sum',
    'profit':'sum'
})
top_five_df

qty  shipping_price   total_price        profit
client_id                                                    
24741      239862      5126448.37  8.226889e+07  3.657999e+07
33615       64313      1828984.89  8.377309e+06  2.201995e+06
38378       73667      3429455.40  1.290655e+07  3.271830e+06
46820       75768      1601448.84  9.743794e+06  2.736603e+06
66037       43018      1395151.85  1.025951e+07  3.255032e+06

In [137]:
# Format the data and rename the columns to names suitable for presentation.
top_five_df = top_five_df.rename(columns={"profit": "Total Profit", "qty": "Quantity", "shipping_price":"Shipping Price", "total_price":"Total Revenue"})
# Define the money columns. 
money_cols = ['Total Profit','Shipping Price', 'Total Revenue']

# Define a function that converts a dollar amount to millions.
def to_million_format(x):
    #Assumption here is that the number is supposed to be stated as for example "1.56 million"
    #Will round to 5 for readability
    return "$" + str(round(x / 1000000, 5)) + " million"

# Apply the currency_format_millions function to only the money columns. 
for col in money_cols:
    top_five_df[col] = top_five_df[col].apply(to_million_format)

# Rename the columns to reflect the change in the money format. 
top_five_df = top_five_df.rename(columns={"Total Price": "Total Price in Millions", "Shipping Price":"Shipping Price in Millions", "Total Revenue": "Total Revenue in Millions", "Total Profit": "Total Profit in Millions"})
top_five_df



Quantity Shipping Price in Millions Total Revenue in Millions  \
client_id                                                                  
24741        239862           $5.12645 million         $82.26889 million   
33615         64313           $1.82898 million          $8.37731 million   
38378         73667           $3.42946 million         $12.90655 million   
46820         75768           $1.60145 million          $9.74379 million   
66037         43018           $1.39515 million         $10.25951 million   

          Total Profit in Millions  
client_id                           
24741            $36.57999 million  
33615             $2.20199 million  
38378             $3.27183 million  
46820              $2.7366 million  
66037             $3.25503 million

In [138]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_top_five = top_five_df.sort_values("Total Profit in Millions", ascending=False)
sorted_top_five

Quantity Shipping Price in Millions Total Revenue in Millions  \
client_id                                                                  
24741        239862           $5.12645 million         $82.26889 million   
38378         73667           $3.42946 million         $12.90655 million   
66037         43018           $1.39515 million         $10.25951 million   
46820         75768           $1.60145 million          $9.74379 million   
33615         64313           $1.82898 million          $8.37731 million   

          Total Profit in Millions  
client_id                           
24741            $36.57999 million  
38378             $3.27183 million  
66037             $3.25503 million  
46820              $2.7366 million  
33615             $2.20199 million

In [154]:
per_quant_df = df[df['client_id'].isin(top_five)].groupby('client_id').agg({
    'qty':'sum',
    'shipping_price':'sum',
    'total_price':'sum',
    'profit':'sum'
})
per_quant_df['profit_per_qty'] = per_quant_df['profit'] / per_quant_df['qty']
print(per_quant_df.sort_values('profit_per_qty', ascending=False))
print(f"The top spender has this job: {df.loc[df['client_id'] == 24741, 'job'].iloc[0]}")
job_check_df = df.loc[df['job'] == "Pension scheme manager",:][['client_id', 'job']].groupby('client_id').count()
print(job_check_df)
# The data shows that the biggest profit earner is client 24741 who is Kendra Garret. This client beats out the other top five spenders by about a 10x difference. 
# This is the case even when the quantity of goods is taken into account which suggests that said client 24741 both buys high volume of items and expensive items. 
# Client 24741 could be an anomalous client, especially given that said client's job is as a 'Pension scheme manager' which is a job shared by only one other client, 
# however consideration of how this client became such a prolific spender could improve sales across the board assuming the spending is not tied solely to the client's salary. 


              qty  shipping_price   total_price        profit  profit_per_qty
client_id                                                                    
24741      239862      5126448.37  8.226889e+07  3.657999e+07      152.504324
66037       43018      1395151.85  1.025951e+07  3.255032e+06       75.666740
38378       73667      3429455.40  1.290655e+07  3.271830e+06       44.413780
46820       75768      1601448.84  9.743794e+06  2.736603e+06       36.118186
33615       64313      1828984.89  8.377309e+06  2.201995e+06       34.238718
The top spender has this job: Pension scheme manager
           job
client_id     
24741      207
49404      142
